### ISSUE TITLE
bismark-methylation-extractor

### ISSUE THREAD

**Issue by kirkupc** [FelixKrueger/Bismark#275]
Hi Felix, 

I've been using Bismark to process some RRBS data, but I'm running into an issue with the bismark-methylation-extractor function. I've used it successfully on some files, but now whenever I run it it produces a full output all the way up to the splitting report, but then doesn't save a .cov file at the end.

I've been navigating to the directory of interest then running the following...
"/home/user1/Bismark-master/bismark_methylation_extractor *.bam"

Which produces this output in the terminal...

```
 Now testing Bismark result file >A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.bam for positional sorting (which would be bad...)	...passed!
Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz

Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz

Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz
Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt.gz

Now reading in Bismark result file A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.bam
skipping SAM header line:	@HD	VN:1.0	SO:unsorted
skipping SAM header line:	@SQ	SN:chr1	LN:248956422
skipping SAM header line:	@SQ	SN:chr10	LN:133797422
skipping SAM header line:	@SQ	SN:chr11	LN:135086622

...


skipping SAM header line:	@SQ	SN:chrY	LN:57227415
skipping SAM header line:	@SQ	SN:chrY_KI270740v1_random	LN:37240
skipping SAM header line:	@PG	ID:Bismark	VN:v0.22.1_dev	CL:"bismark --path_to_bowtie2 /home/user1/anaconda2/bin /home/user1/Bismark-master/genomes -1 A3_S11_L001_R1_001.fastq_trimmed.fq.gz -2 A3_S11_L001_R2_001.fastq_trimmed.fq.gz"
Processed lines: 500000
Processed lines: 1000000
Processed lines: 1500000
Processed lines: 2000000
Processed lines: 2500000
Processed lines: 3000000
Processed lines: 3500000
Processed lines: 4000000
Processed lines: 4500000
Processed lines: 5000000
Processed lines: 5500000
Processed lines: 6000000
Processed lines: 6500000
Processed lines: 7000000

Processed 7248549 lines in total
Total number of methylation call strings processed: 14497098

Final Cytosine Methylation Report
=================================
Total number of C's analysed:	248142424

Total methylated C's in CpG context:	19361902
Total methylated C's in CHG context:	363932
Total methylated C's in CHH context:	581372

Total C to T conversions in CpG context:	17422279
Total C to T conversions in CHG context:	64355908
Total C to T conversions in CHH context:	146057031

C methylated in CpG context:	52.6%
C methylated in CHG context:	0.6%
C methylated in CHH context:	0.4%


Determining maximum read lengths for M-Bias plots
Maximum read length of Read 1: 74
Maximum read length of Read 2: 71

Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)
Determining maximum read lengths for M-Bias plots
Maximum read length of Read 1: 74
Maximum read length of Read 2: 71

Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)
Deleting unused files ...

CpG_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt contains data ->	kept
CpG_CTOT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt was empty ->	deleted
CpG_CTOB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt was empty ->	deleted
CpG_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt contains data ->	kept
CHG_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt contains data ->	kept
CHG_CTOT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt was empty ->	deleted
CHG_CTOB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt was empty ->	deleted
CHG_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt contains data ->	kept
CHH_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt contains data ->	kept
CHH_CTOT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt was empty ->	deleted
CHH_CTOB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt was empty ->	deleted
CHH_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt contains data ->	kept
```
The directory contains these files after running the command...

```
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe_splitting_report.txt
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.M-bias.txt
/home/user1/RBBS/KP50503_3_postQC/CpG_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt
/home/user1/RBBS/KP50503_3_postQC/CpG_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt
/home/user1/RBBS/KP50503_3_postQC/CHH_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt
/home/user1/RBBS/KP50503_3_postQC/CHH_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt
/home/user1/RBBS/KP50503_3_postQC/CHG_OT_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt
/home/user1/RBBS/KP50503_3_postQC/CHG_OB_A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.txt
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_PE_report.txt
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R1_001.fastq_trimmed_bismark_bt2_pe.bam
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R2_001.fastq_trimmed.fq.gz
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R1_001.fastq_trimmed.fq.gz
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R2_001.fastq.gz
/home/user1/RBBS/KP50503_3_postQC/A3_S11_L001_R1_001.fastq.gz
```

Any suggestions you might have on how to resolve my issue would be much appreciated. Thanks!

**Comment by FelixKrueger** [comment]
Did you specify the command `--bedGraph` when running the methylation extractor? That option will kick off the `bismark2bedGraph` which generates a bedGraph and coverage file. You could in theory also run `bismark2bedGraph` on its own, using e.g. just the CpG* files, like so:

```
bismark2bedGraph -o outputfile.name CpG*
```

I am now on holiday and might not be able to answer for some time, hope you'll find the issue.

**Comment by kirkupc** [comment]
Hi Felix, 

My mistake, I was under the impression that the .cov file was generated regardless of if the BedGraph option was enabled. That solves my issue. 

Thanks, have a good holiday!

**Comment by sangeeta-25** [comment]
I have a query regarding same issue. I have run this command for extraction:
bismark_methylation_extractor --gzip --bedGraph --buffer_size 10G sample_bismark_bt2.bam
I got the result file mention above but C15_CpG_report.txt.gz amd  text coverage file is empty. I could understand how could I tackle this  problem. please help me through this. Thank you in advance.

**Comment by FelixKrueger** [comment]
Did you some hints of what might have gone wrong from the log output (or on screen)?

Does your genome have many scaffolds/contigs? In that case you might want to look into the option `--scaffolds`. (or `--help` to see the help text

**Comment by sangeeta-25** [comment]
My genome having contigs but my query is why i am not getting the .cov file

**Comment by FelixKrueger** [comment]
The reason is given in the error message on screen, and also in help text. Allow me to link it here again:

```
--scaffolds/--gazillion  Users working with unfinished genomes sporting tens or even hundreds of thousands of
                         scaffolds/contigs/chromosomes frequently encountered errors with pre-sorting reads to
                         individual chromosome files. These errors were caused by the operating system's limit
                         of the number of filehandle that can be written to at any one time (typically 1024; to
                         find out this limit on Linux, type: ulimit -a).
                         To bypass the limitation of open filehandles, the option --scaffolds does not pre-sort
                         methylation calls into individual chromosome files. Instead, all input files are
                         temporarily merged into a single file (unless there is only a single file), and this
                         file will then be sorted by both chromosome AND position using the Unix sort command.
                         Please be aware that this option might take a looooong time to complete, depending on
                         the size of the input files, and the memory you allocate to this process (see --buffer_size).
                         Nevertheless, it seems to be working
```

**Comment by sangeeta-25** [comment]
Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table) = also not gettinh the M-bias plot. is there any alternative tool through which i can plot the graph.
I am getting this on terminal but file is empty:

Writing bedGraph to file: C15_L1_R1_trimmed_bismark_bt2_pe.bedGraph.gz
Also writing out a coverage file including counts methylated and unmethylated residues to file: C15_L1_R1_trimmed_bismark_bt2_pe.bismark.cov.gz

Failed to open filehandle: Too many open files at /home/user/Downloads/Bismark-0.24.2/bismark2bedGraph line 279, <IN> line 237895.
Finished BedGraph conversion ...
Above is the last line i got on the terminal
Kindly give your valuable comments

**Comment by FelixKrueger** [comment]
You could either install the GD::Graph module, or alternatively you can run `bismark2report` once methylation extractor is complete to create an interactive HTML report (which also contains the M-bias plot).

Exactly, this error message:

```
Failed to open filehandle: Too many open files at
```

is the reason why you need to use the `--scaffolds`. Just use it, and all will be well...

**Comment by sangeeta-25** [comment]
Inside the bismark_methylation_extractor module in bismark there is two option for extraction one is normal and another is genome . can anyone explain the main purpose of these two different command option? How it is going to effect the downstream analysis of epigenome data.

**Comment by FelixKrueger** [comment]
The extractions procedure itself doesn't discriminate between _normal_ and _genome_; you may optionally transform the the primary methylation extractor output further to bedGraph/coverage format, and further to genome-wide cytosine reports (maybe this is what you were referring to?)? Details about this can be found in the documentation here: https://felixkrueger.github.io/Bismark/bismark/methylation_extraction/

**Comment by FelixKrueger** [comment]
I think the answer to your question is 'no', but there could be a confusion with terminology. In essence, the methylation extractor (step 1 below) runs always, steps 2 (`--bedGraph`) and 3 (`--cytososine_report` for which you need to specify `--genome ...`) are optional.

1. the extraction produces CpG, CHG and CHH context files, as well as an `splitting report` 
2. if `--bedGraph` is selected, the files produced in 1 are sorted and counted in bedGraph and coverage files
3. if `--cytosine_report` is selected, the coverage file from 2. is used as input, related back to the `genome` (that needs to be speecified, resulting in genome-wide cytosine reports

I hope this clarifies things?

**Comment by sangeeta-25** [comment]
Thank you so much . Now i got the answe.

**Comment by sangeeta-25** [comment]
Can you suggest me the downstream analysis tool to identify DMRs in case of plant. I have used the methylkit for the analysis but somehow i am not satisfied with the result. I am also using the Batmeth2 but it is too slow. If anyone can suggest.

**Comment by FelixKrueger** [comment]
I can also recommend Seqmonk for these kinds of things, have a look at the training material available here for a flavour of what you can do: https://www.bioinformatics.babraham.ac.uk/training.html#bsseq

**Comment by sangeeta-25** [comment]
chr start end strand pvalue qvalue meth.diff
Chr4 29864211 29864211 * 1.54279819171964e-06 0.00724512056269648
-33.0236486486487
Chr7 49209664 49209664 * 3.36164577193072e-12 1.30239401730469e-07
-45.9381443298969
Chr7 49209674 49209674 * 1.74615170495659e-11 4.51004793114714e-07
-46.7762011451332
Contig01936 6267 6267 * 1.38877173300439e-22 2.15219344235423e-17
-33.6555855651194
Contig01936 6273 6273 * 1.43999548282857e-15 7.43858466564755e-11
-29.5401945724526
Contig01936 6281 6281 * 2.22972725687242e-16 1.72771531362388e-11
-31.2617702448211
Contig01936 6281 6281 * 7.03211527617433e-12 2.17954787292802e-07
25.7492564630519
Contig01936 6300 6300 * 4.88880520381458e-07 0.0039874896381071
31.6076085289155


Can one chromosome have hypermethylation and hypomethylation at the same
position as mentioned in the above result? Please help me through this. i
will be thankful.

**Comment by FelixKrueger** [comment]
You should probably ask this question on the repository of the program you used to arrive at these numbers... 

```
Contig01936      6281     -31.2617702448211
Contig01936      6281      25.7492564630519
```

The only potential situation I can think of where you could have different values for the very same position would be if the tool you used to arrive at these numbers discriminates between different methylation contexts. Such [context changes](https://felixkrueger.github.io/Bismark/faq/context_change/) may occasionally happen for positions that have a deletion right next to a cytosine position. In theory, there might exist positions with and without a deletion on different alleles... Please note that this is very theoretical at this point....

I think you should check what exactly you did to get the results above.

**Comment by FelixKrueger** [comment]
I really don't have enough details to judge what you did. If you have questions regarding metylKit you should talk to al2na & team.
